### PR TITLE
Business rule task can evaluate a DMN decision

### LINKED ISSUE #8080: A business rule task can evaluate a DMN decision
**Description**

A business rule task can reference a DMN decision by its decision id. If a process instance activates such a business rule task then it evaluates the decision. It writes the result of the evaluation as a process variable under the name that is defined as the result variable. The value type of the variable depends on the DMN decision (i.e. single value, list, or context).

* extend the executable representation of a business rule task
  * add a property for the decision id
  * add a property for the result variable
* evaluate the decision when processing the `activate` command
  * after the input mappings are applied
  * load the decision from the state
  * use the DMN module to evaluate the decision
* write the decision result as process variable
  * the decision result is returned by the DMN module
  * the name of the variable is defined as `resultVariable`
  * propagate the variable if no output mappings are defined
  * set the variable in the local scope if output mappings are defined (e.g. to extract or transform the decision result)
  * use the process event to pass the decision result to the output mapping

blocked by #8060 
blocked by #8067 
blocked by #8069
blocked by #8082

### PR DESCRIPTION
**Pull request by korthout**
## Description

This adds the ability to process business rule tasks that refer to a called decision (see #8060). Business rule tasks that refer to a [task definition](https://docs.camunda.io/docs/components/modeler/bpmn/business-rule-tasks/business-rule-tasks/#defining-a-task) instead of a called decision work like they did before, i.e. by creating a job for job workers.

On activation of a business rule task with called decision, the decision specified by `decisionId` is retrieved from the deployed decisions state. If present, the engine uses the `dmn` module to evaluate the decision. If that is successful, the decision result's output is stored in a result variable (defined by `resultVariable`).

The result variable stores the decision result's output according to the following rules:
- if the business rule task defines no output mappings, the result variable is [propagated](https://docs.camunda.io/docs/components/concepts/variables/#variable-propagation)
- if there are [output mappings](https://docs.camunda.io/docs/components/concepts/variables/#output-mappings) defined on the business rule task, the result variable is set as a [local variable](https://docs.camunda.io/docs/components/concepts/variables/#local-variables) of the business rule task

If any of this fails, Zeebe raises an incident on the business rule task. For example, if the decision was not deployed yet, then an incident is raised. This allows you to:
1. deploy the decision to fix the problem
2. retry activating the business rule task by resolving the incident

## Related issues

closes #8080

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

### CODE REVIEW

**Comment by korthout** [comment]
🤔 It looks like the `OS Smoke Tests` and the `Code QL` fail for the same reason:
```
[INFO] -------------------------------------------------------------
Error:  COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
Error:  /home/runner/work/zeebe/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java:[13,28] cannot find symbol
  symbol:   class DecisionResult
  location: package io.camunda.zeebe.dmn
Error:  /home/runner/work/zeebe/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java:[59,26] cannot find symbol
  symbol:   class DecisionResult
  location: class io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnDecisionBehavior
Error:  /home/runner/work/zeebe/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java:[123,27] cannot find symbol
  symbol:   class DecisionResult
  location: class io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnDecisionBehavior
Error:  /home/runner/work/zeebe/zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnDecisionBehavior.java:[138,13] cannot find symbol
  symbol:   class DecisionResult
  location: class io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnDecisionBehavior
[INFO] 4 errors 
```

It looks like it is able to find the other classes from `io.camunda.zeebe.dmn`, but just not the `DecisionResult`. Perhaps this is a concurrency problem 🤔 . All those tests run: `mvn ... -T1C`, so perhaps the dmn jar is not installed locally before it tries to compile the workflow-engine. I'll have another look tomorrow.

**Comment by korthout** [comment]
Manually tested in Camunda Cloud SaaS (by using the QA run to let testbench create a new tmp generation based on this code and quickly using this generation to create a personal cluster). 

Here's a screenshot of a business rule task with called decision that calls the `jedi_or_sith` decision from the `drg-force-user.dmn` test resource. It requires a `lightsaberColor` variable. Its result variable is set to `result`.
Screen Shot 2022-02-11 at 14 22 32

Next to this, I've also manually tested:
- using a decision id for which no decisions are deployed:
  - correctly produced an incident
- not passing in one of the required variables:
  - correctly produced an incident
  - incident could be resolved after adding the variable to the process instance
- using an input mapping
  - correctly passes the input mapping's result to the decision
- using an output mapping 
  - correctly makes the result variable into a local variable
  - correctly uses the local result variable in the output mapping

**Review by saig0 — APPROVED** [review]
@korthout well done :1st_place_medal: 

The code looks simple and clean. Thanks for splitting the PR into the commits and adding the descriptions :+1: 

I just have some minor suggestions. Feel free to merge :rocket:

File: util/src/main/java/io/camunda/zeebe/util/Either.java [hunk 1]
```diff
@@ -456,4 +480,18 @@ public String toString() {
       return "Left[" + value + ']';
     }
   }
+
+  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+  final class EitherOptional<R> {
+
+    private final Optional<R> right;
+
+    public EitherOptional(final Optional<R> right) {
```
[the remark this answers is not in the document]
Comment (korthout) [reply util/src/main/java/io/camunda/zeebe/util/Either.java:489]: @saig0 We also turn it into a `record`.

```java
record EitherOptional<R>(Optional<R> right) {
  public <L> Either<L, R> orElse(final L left) {
    return right.<Either<L, R>>map(Either::right).orElse(Either.left(left));
  }
}
```

In this case, you have 2 methods available after using `Either.ofOptional`:
Screen Shot 2022-02-14 at 15 47 47

WDYT?

**Comment by korthout** [comment]
bors merge

**Comment by korthout** [comment]
bors retry

**Comment by korthout** [comment]
> Build failed:
> 
>     * [continuous-integration/jenkins/branch](https://ci.zeebe.camunda.cloud/job/camunda-cloud/job/zeebe/job/staging/1819/display/redirect)

Ugh, aborted due to connection error.

bors merge

**Pull request merged**